[#56] feat : 위시리스트(CART) 추가/삭제

kihyuk-jeong opened the pull request:
1. 카트에 Product 추가
2. 카트에서 Product 제거

 #56 : 위시리스트(CART) 추가/삭제

kihyuk-jeong committed:
[#54] feat : 위시리스트 도메인 설계

1. Cart
2. CartProduct
3. User

 #54 : 위시리스트(CART) 모델링

kihyuk-jeong committed:
[#56] feat : 위시리스트(CART) 추가/삭제

1. 카트에 Product 추가
2. 카트에서 Product 제거

 #56 : 위시리스트(CART) 추가/삭제

kihyuk-jeong committed:
chore : UaserLevel 기본값 수정

1. 테스트를 위해 변경했던 기본값 UNAUTH로 수정

kihyuk-jeong committed:
chore : 핸들러 접근 권한 수정

1. Cart 관련 핸들러 @LoginCheck 어노테이션 추가

Hosick commented on src/main/java/com/flab/shoeauction/controller/dto/ProductDto.java:
```diff
@@ -107,4 +108,37 @@ public Product toEntity() {
+        @Getter
+        @NoArgsConstructor
+        public static class WishItemResponse {
+
```
RequestDto 안에 ResponseDto 스태틱 클래스가 존재하는 설계인가요?
어떻게 사용되는지 감이 잘 안오는데 혹시 설명 부탁드려도 될까요?

kihyuk-jeong replied:
이부분은 구현하는 과정에서 실수가 있었던 것 같습니다 😅  정상적으로 수정 했고 원래 사용하던 방식대로 생각하시면 될 것 같습니다!

Hosick commented on src/main/java/com/flab/shoeauction/exception/GlobalExceptionHandler.java:
```diff
@@ -160,4 +161,11 @@ public final ResponseEntity handleNotAuthorizedException(NotAuthorizedException
+    @ExceptionHandler(EmptyCartException.class)
+    public final ResponseEntity<Void> handleEmptyCartException (EmptyCartException ex) {
+        log.debug("위시리스트가 비어있습니다.", ex);
+        return BAD_REQUEST;
```
@f-lab-dev 이 부분에서 궁금한 점이 있습니다!

 토비의 스프링을 읽고 이해한 것으로는 `예외 처리`는 개발자가 생각한 정상적인 로직 흐름에서 벗어났을 때 실시해야 한다고 알고 있습니다. 그렇다면 이렇게 전체조회에서 0건이 발생한 경우에도 정상적인 흐름을 벗어났다고 봐야할까요?

f-lab-dev replied:
이거는 예외 상황이 아닌 것 같습니다~ 장바구니가 비어있을 수 있죠

kihyuk-jeong replied:
제가 생각했을때는 비어있는 장바구니를 예외라고 생각했었는데 다시생각해보니 정상적인 로직에서 벗어나는 흐름은 아닌 것 같습니다. 빈 장바구니도 조회될 수 있도록 수정하겠습니다~

kihyuk-jeong committed:
refactor : 리뷰 반영

1. ProductDto 수정

f-lab-dev commented on src/main/java/com/flab/shoeauction/controller/UserApiController.java:
```diff
@@ -193,4 +195,23 @@ public void updateAddressBook(@RequestBody AddressBookDto requestDto) {
     }
+
+    @LoginCheck
+    @GetMapping("/cart")
```
이런 URL은 복수형으로 해주는게 좋지 않을까요?

kihyuk-jeong committed:
refactor : 리뷰 반영

1. 비어있는 Cart 조회시 예외처리가 아닌 비어있는 컬렉션 리턴하도록 수정

kihyuk-jeong committed:
refactor : 리뷰 반영 + 장바구니 중복 검사

1. cart URL 복수형으로 수정
2. 장바구니에 동일한 아이템을 추가할 경우 예외처리
3. 디미터 법칙에 위배되는 메소드 수정

 #56 : 위시리스트(CART) 추가/삭제

kihyuk-jeong committed:
chore : 롬복 수정

1. @EqualsAndHashCode 삭제

kihyuk-jeong commented:
- 장바구니 중복 검사 로직 추가
- 디미터 법칙에 위배되는 메소드 수정

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/utils/constants/ResponseConstants.java:
```diff
@@ -65,4 +65,9 @@
             HttpStatus.BAD_REQUEST);
+
+    public static final ResponseEntity<String> DUPLICATION_CART_ITEM = new ResponseEntity<>(
+        "해당 상품은 이미 위시리스트에 등록되어 있는 상품입니다.",
```
장바구니에 같은 상품을 여러개 추가할 수도 있을 것 같은데요~ 중복으로 넣으면 수량을 올려준다거나 하면 되지 않을까요?
물론 기획에 따라 다르긴 할 것 같네요

kihyuk-jeong replied:
@f-lab-dev
네. 이 부분은 말씀하신 것처럼 프로젝트의 특성상 장바구니에 수량 개념을 사용하지 않았습니다.
일반 쇼핑몰과 다르게 경매 시스템을 근본으로 하여 상품을 구매하는 시스템이기 때문에 수량을 많이 추가한다고 해도 수량마다 판매자가 모두 달라서 일괄 구매가 불가능합니다.
따라서 해당 프로젝트에서는 `북마크`의 개념으로 장바구니가 사용될 것 같습니다.!
